pkg/kv/kvserver/flow_control_stores.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package kvserver
+
+import (
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+)
+
+// StoresForFlowControl is used to integrate with replication flow control. It
+// exposes the underlying kvflowcontrol.Handles and is informed of (remote)
+// stores we're no longer connected via the raft transport.
+type StoresForFlowControl interface {
+	kvflowcontrol.Handles
+	RaftTransportDisconnectListener
+}
+
+// storesForFlowControl is a concrete implementation of the
+// StoresForFlowControl interface, backed by a set of Stores.
+type storesForFlowControl Stores
+
+var _ StoresForFlowControl = &storesForFlowControl{}
+
+// MakeStoresForFlowControl returns the canonical StoresForFlowControl
+// implementation.
+func MakeStoresForFlowControl(stores *Stores) StoresForFlowControl {
+	return (*storesForFlowControl)(stores)
+}
+
+// Lookup is part of the StoresForFlowControl interface.
+func (sh *storesForFlowControl) Lookup(
+	rangeID roachpb.RangeID,
+) (handle kvflowcontrol.Handle, found bool) {
+	ls := (*Stores)(sh)
+	if err := ls.VisitStores(func(s *Store) error {
+		if h, ok := makeStoreForFlowControl(s).Lookup(rangeID); ok {
+			handle = h
+			found = true
+		}
+		return nil
+	}); err != nil {
+		ctx := ls.AnnotateCtx(context.Background())
+		log.Errorf(ctx, "unexpected error: %s", err)
+		return nil, false
+	}
+	return handle, found
+}
+
+// ResetStreams is part of the StoresForFlowControl interface.
+func (sh *storesForFlowControl) ResetStreams(ctx context.Context) {
+	ls := (*Stores)(sh)
+	if err := ls.VisitStores(func(s *Store) error {
+		makeStoreForFlowControl(s).ResetStreams(ctx)
+		return nil
+	}); err != nil {
+		ctx = ls.AnnotateCtx(ctx)
+		log.Errorf(ctx, "unexpected error: %s", err)
+	}
+}
+
+// OnRaftTransportDisconnected is part of the StoresForFlowControl
+// interface.
+func (sh *storesForFlowControl) OnRaftTransportDisconnected(
+	ctx context.Context, storeIDs ...roachpb.StoreID,
+) {
+	ls := (*Stores)(sh)
+	if err := ls.VisitStores(func(s *Store) error {
+		makeStoreForFlowControl(s).OnRaftTransportDisconnected(ctx, storeIDs...)
+		return nil
+	}); err != nil {
+		ctx := ls.AnnotateCtx(context.Background())
+		log.Errorf(ctx, "unexpected error: %s", err)
+	}
+}
+
+// storeForFlowControl is a concrete implementation of the
+// StoresForFlowControl interface, backed by a single Store.
+type storeForFlowControl Store
+
+var _ StoresForFlowControl = &storeForFlowControl{}
```

Comment:
Is it necessary that both multi-store and per-store implementation share the same interface? Does the consumer of this always use the `Stores` (the multi-store impl) to gate the requests, and the per-store struct is just an implementation detail?

Reply:
> Is it necessary that both multi-store and per-store implementation share the same interface?

It's not necessary, no. Just a code organization opinion (and possibly a bad one). I'll remove if you tell me to, though the methods would be very similarly named.

> Does the consumer of this always use the Stores (the multi-store impl) to gate the requests, and the per-store struct is just an implementation detail?

It uses the multi-store implementation.